File: src/tasks/topic_modeling.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import logging
from typing import List, Optional
from src.models.genai_model import GenaiModel
from src import sensemaker_utils
from src.tasks import topic_modeling_util
from src.models import custom_types


LEARN_TOPICS_PROMPT = """
You are an expert qualitative data analyst specializing in thematic analysis and data structuring.
Your task is to analyze this entire data,
identify the topics according to the criteria below,
and then generate a JSON output with the identified topics.

Important Context:
There will be another round, where participants will be exploring topics, so it will be easier for them to tie everything together in their head, if they see topics being linked to the main subject.
To facilitate this connection, we want topic language to more explicitly connect to the main subject. Topics should be phrased like they are aspects of the main subject.

### **Criteria for Topics**

  * **Distinct:** Topics should be meaningfully different and cover separate conceptual areas.
  * **Substantive:** Topics should contain multiple, distinct opinions. Do not create single-opinion topics.
  * **Subject Linkage:** Topic names should be phrased as aspects of the main subject. E.g. "Freedom and Equality" subject could have the following topics: "Defining Freedom", "Defining Equality", "Barriers to Freedom and Equality", "Society's Role in Freedom and Equality", "The Individual's Role in Freedom and Equality", etc.
  * **Concise:** The topic name should be concise.
  * **Consistent Scope:** Topics should be at a similar level of abstraction (e.g., don't mix a very broad topic with a very narrow one).
  * **Efficiency:** Keep the number of topics as low as possible. Actively consolidate topics when their content can be logically grouped.


RESPONSE STRUCTURE:
Respond with a list of the identified topics only, nothing else.
The response should be in JSON format, that can be parse into the following class:
class FlatTopicList:
    topics: List[Topic]
class Topic:
    name: str

Do not include markdown code blocks around the JSON response, such as ```json or ```
Response example:
{"topics": [{"name": "Topic 1"}, {"name": "Topic 2"}]}
"""


def learn_opinions_prompt(parent_topic: custom_types.Topic) -> str:
  return f"""
You are an expert qualitative data analyst specializing in thematic analysis and data structuring.
Your task is to analyze this entire dataset of quotes, identify the opinions on the following topic: "{parent_topic.name}" according to the criteria below, and then generate a JSON output with the identified opinions.

### **Criteria for Opinions**

1.  **Active Voice & Direct Phrasing (Crucial):**
    * Use strong, active verbs. Avoid passive voice (e.g., "It is believed that...").
    * Avoid abstract policy speak. Instead of "To improve economic opportunity, there needs to be investment in...", write "Oklahoma must invest in..." or "Schools need better funding to..."
    * **Do not** use words like "perception" or "sentiment." State the opinion as a fact as viewed by the participant.

2.  **Avoid Repetitive Sentence Starters:**
    * **Do not** start every opinion with the same phrase (e.g., stop using "To strengthen the social safety net..." for every single item).
    * Ensure the list has syntactic variety while remaining thematically tight.

3.  **Simplify & Avoid Complex Parallelisms:**
    * **One idea per opinion.** Avoid complex lists (e.g., "We need X, Y, and Z, while also ensuring A and B").
    * Aim for a 5th-grade reading level. Keep it simple and punchy.

4.  **Distinct & Substantive:**
    * Opinions must represent unique viewpoints within the topic.
    * Merge overlaps: Actively consolidate opinions when their content can be logically grouped.
    * Do not create single-quote opinions or opinions with very few opinions compared to its peers. A long tail of opinions is extremely undesirable.
    * Overall, we want to tightly curate opinions to help the user understand the main perspective within a topic, but we do not want to overwhelm them with a laundry list.  **Keep the number of opinions as low as possible.**

5.  **Topic Linkage (Without Repetition):**
    * The opinion must be clearly relevant to "{parent_topic.name}", but it should not rigidly repeat the topic name in the text.
    * *Bad:* "A barrier to economic growth is the lack of jobs."
    * *Good:* "A lack of quality jobs prevents the economy from growing."

### **Response Structure**
Respond only with the identified opinions, where top level is the overarching topic, and opinions are subtopics.
The response should be in JSON format, that can be parse into the following class:
class Topic:
    name: str # This will be the overarching topic
    subtopics: List[Topic] # Where subtopic is class Topic {{ name: str }} (the opinions)

Do not include markdown code blocks around the JSON response, such as ```json or ```
For example:
{{
  "name": "{parent_topic.name}",
  "subtopics": [
      {{ "name": "Opinion 1" }},
      {{ "name": "Opinion 2" }}
  ]
}}
"""


def _are_valid_topics(response: List[custom_types.FlatTopic], *args) -> bool:
  """Validates that the response is a list of FlatTopic."""
  if not isinstance(response, list):
    logging.warning(
        f"Validation failed: Response is not a list. Got: {type(response)}"
    )
    return False
  if not all(isinstance(t, custom_types.FlatTopic) for t in response):
    logging.warning(
        f"Validation failed: Not all topics are FlatTopic. Response: {response}"
    )
    return False
  return True


def _is_valid_opinion(
    response: Optional[custom_types.NestedTopic],
    parent_topic: custom_types.Topic,
) -> bool:
  """Validates the structure for a learned opinions response."""
  if not response or not isinstance(response, custom_types.NestedTopic):
    logging.warning(
        "Validation failed: Response is not a NestedTopic. Got:"
        f" {type(response)}"
    )
    return False
  # When learning opinions, expect a single nested topic with the parent's name.
  if response.name != parent_topic.name:
    logging.warning(
        f"Validation failed for sub-level topics of '{parent_topic.name}'."
        f" Response: {response}"
    )
    return False
  return True


async def learn_topics(
    statements: list[custom_types.Statement],
    model: GenaiModel,
    additional_context: Optional[str] = None,
) -> list[custom_types.FlatTopic]:
  """Learns top-level topics from a list of statements."""
  instructions = LEARN_TOPICS_PROMPT
  schema_to_expect = custom_types.FlatTopicList
  logging.debug("Using LEARN_TOPICS_PROMPT (expecting FlatTopicList)")

  prompt_input_data = [statement_item.text for statement_item in statements]

  if not prompt_input_data:
    logging.warning(
        "No statements provided to learn topics from. Returning empty list."
    )
    return []

  chunks = await topic_modeling_util.create_chunks(
      model, instructions, prompt_input_data, additional_context
  )

  async def _generate_topics(
      current_model: GenaiModel,
  ) -> List[custom_types.FlatTopic]:
    logging.info(
        f"Identifying topics for {len(prompt_input_data)} input items..."
    )
    result = await topic_modeling_util.generate_topics_with_chunking(
        model=current_model,
        instructions=instructions,
        prompt_input_data=prompt_input_data,
        schema_to_expect=schema_to_expect,
        additional_context=additional_context,
        chunks=chunks,
    )
    if isinstance(result, custom_types.FlatTopicList):
      return result.topics
    return result

  response_topics = await sensemaker_utils.retry_call(
      _generate_topics,
      _are_valid_topics,
      model.max_llm_retries,
      "Topic identification failed after multiple retries.",
      func_args=[model],
      is_valid_args=[],
  )

  if response_topics:
    return response_topics

  logging.error(
      "Could not generate topics after retries. Returning empty list."
  )
  return []


async def learn_opinions(
    statements: List[custom_types.Statement],
    model: GenaiModel,
    topic: custom_types.Topic,
    additional_context: Optional[str] = None,
) -> custom_types.NestedTopic:
  """Learns opinions (as subtopics) for a given parent topic."""
  prompt_input_data: List[str] = []
  for statement_item in statements:
    if statement_item.quotes:
      relevant_quotes_for_statement = [
          q for q in statement_item.quotes if q.topic.name == topic.name
      ]
      for quote_obj in relevant_quotes_for_statement:
        prompt_input_data.append(f"<quote>{quote_obj.text}</quote>")

  if not prompt_input_data:
    logging.warning(
        f"No relevant quotes for topic '{topic.name}' to generate opinions"
        " from. Returning parent topic with empty subtopics."
    )
    return custom_types.NestedTopic(name=topic.name, subtopics=[])

  instructions = learn_opinions_prompt(topic)
  # Use non-recursive schema to avoid RecursionError in SDK
  schema_to_expect = custom_types.OpinionResponseSchema
  logging.debug(
      f"Using learn_opinions_prompt for topic: {topic.name} (expecting"
      " OpinionResponseSchema)"
  )

  chunks = await topic_modeling_util.create_chunks(
      model, instructions, prompt_input_data, additional_context
  )

  async def _generate_opinions(
      current_model: GenaiModel,
  ) -> Optional[custom_types.NestedTopic]:
    logging.info(
        f"Identifying opinions for topic '{topic.name}' from"
        f" {len(prompt_input_data)} quotes..."
    )

    result = await topic_modeling_util.generate_opinions_with_chunking(
        model=current_model,
        instructions=instructions,
        prompt_input_data=prompt_input_data,
        schema_to_expect=schema_to_expect,
        parent_topic=topic,
        additional_context=additional_context,
        chunks=chunks,
    )

    if isinstance(result, custom_types.OpinionResponseSchema):
      # Convert back to NestedTopic
      return custom_types.NestedTopic(
          name=result.name, subtopics=result.subtopics
      )
    return result

  response_topic = await sensemaker_utils.retry_call(
      _generate_opinions,
      _is_valid_opinion,
      model.max_llm_retries,
      "Opinion identification or restructuring failed after multiple retries.",
      func_args=[model],
      is_valid_args=[topic],
  )

  if response_topic:
    return response_topic

  logging.error(
      f"Could not generate opinions for topic '{topic.name}' after"
      " retries. Returning parent with empty subtopics."
  )
  return custom_types.NestedTopic(name=topic.name, subtopics=[])

File: case_studies/wcbgb/library/src/tasks/topic_modeling.ts
// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { Type } from "@sinclair/typebox";
import { Model } from "../models/model";
import { MAX_RETRIES } from "../models/model_util";
import { getPrompt, retryCall } from "../sensemaker_utils";
import { Comment, FlatTopic, NestedTopic, Topic } from "../types";

/**
 * @fileoverview Helper functions for performing topic modeling on sets of comments.
 */

export const LEARN_TOPICS_PROMPT = `
Analyze the following comments and identify common topics.
Consider the granularity of topics: too few topics may oversimplify the content and miss important nuances, while too many topics may lead to redundancy and make the overall structure less clear.
Aim for a balanced number of topics that effectively summarizes the key themes without excessive detail.
After analysis of the comments, determine the optimal number of topics to represent the content effectively.
Justify why having fewer topics would be less optimal (potentially oversimplifying and missing key nuances), and why having more topics would also be less optimal (potentially leading to redundancy and a less clear overall structure).
After determining the optimal number of topics, identify those topics.
`;

export function learnSubtopicsForOneTopicPrompt(parentTopic: Topic, otherTopics?: Topic[]): string {
  const otherTopicNames = otherTopics?.map((topic) => topic.name).join(", ") ?? "";

  return `
Analyze the following comments and identify common subtopics within the following overarching topic: "${parentTopic.name}".
Consider the granularity of subtopics: too few subtopics may oversimplify the content and miss important nuances, while too many subtopics may lead to redundancy and make the overall structure less clear.
Aim for a balanced number of subtopics that effectively summarizes the key themes without excessive detail.
After analysis of the comments, determine the optimal number of subtopics to represent the content effectively.
Justify why having fewer subtopics would be less optimal (potentially oversimplifying and missing key nuances), and why having more subtopics would also be less optimal (potentially leading to redundancy and a less clear overall structure).
After determining the optimal number of subtopics, identify those subtopics.

Important Considerations:
- No subtopics should have the same name as the overarching topic.
- There are other overarching topics that are being used on different sets of comments, do not use these overarching topic names as identified subtopics names: ${otherTopicNames}

Example of Incorrect Output:

[
  {
    "name": "Economic Development",
    "subtopics": [
        { "name": "Job Creation" },
        { "name": "Business Growth" },
        { "name": "Small Business Development" },
        { "name": "Small Business Marketing" } // Incorrect: Too closely related to the "Small Business Development" subtopic
        { "name": "Infrastructure & Transportation" } // Incorrect: This is the name of a main topic
      ]
  }
]
`;
}

/**
 * Generates an LLM prompt for topic modeling of a set of comments.
 *
 * @param parentTopics - Optional. An array of top-level topics to use.
 * @returns The generated prompt string.
 */
export function generateTopicModelingPrompt(parentTopic?: Topic, otherTopics?: Topic[]): string {
  if (parentTopic) {
    return learnSubtopicsForOneTopicPrompt(parentTopic, otherTopics);
  } else {
    return LEARN_TOPICS_PROMPT;
  }
}

/**
 * Learn either topics or subtopics from the given comments.
 * @param comments the comments to consider
 * @param model the LLM to use
 * @param topic given or learned topic that subtopics will fit under
 * @param otherTopics other topics that are being used, this is used
 * to avoid duplicate topic/subtopic names
 * @param additionalContext more info to give the model
 * @returns the topics that are present in the comments.
 */
export function learnOneLevelOfTopics(
  comments: Comment[],
  model: Model,
  topic?: Topic,
  otherTopics?: Topic[],
  additionalContext?: string
): Promise<Topic[]> {
  const instructions = generateTopicModelingPrompt(topic, otherTopics);
  const schema = topic ? Type.Array(NestedTopic) : Type.Array(FlatTopic);

  return retryCall(
    async function (model: Model): Promise<Topic[]> {
      console.log(`Identifying topics for ${comments.length} statements`);
      return (await model.generateData(
        getPrompt(
          instructions,
          comments.map((comment) => comment.text),
          additionalContext
        ),
        schema
      )) as Topic[];
    },
    function (response: Topic[]): boolean {
      return learnedTopicsValid(response, topic);
    },
    MAX_RETRIES,
    "Topic identification failed.",
    undefined,
    [model],
    []
  );
}

/**
 * Validates the topic modeling response from the LLM.
 *
 * @param response The topic modeling response from the LLM.
 * @param parentTopics Optional. An array of parent topic names to validate against.
 * @returns True if the response is valid, false otherwise.
 */
export function learnedTopicsValid(response: Topic[], parentTopic?: Topic): boolean {
  const topicNames = response.map((topic) => topic.name);

  // 1. If a parentTopic is provided, ensure no other top-level topics exist except "Other".
  if (parentTopic) {
    const allowedTopicNames = [parentTopic]
      .map((topic: Topic) => topic.name.toLowerCase())
      .concat("other");
    if (!topicNames.every((name) => allowedTopicNames.includes(name.toLowerCase()))) {
      topicNames.forEach((topicName: string) => {
        if (!allowedTopicNames.includes(topicName.toLowerCase())) {
          console.warn(
            "Invalid response: Found top-level topic not present in the provided topics. Provided topics: ",
            allowedTopicNames,
            " Found topic: ",
            topicName
          );
        }
      });
      return false;
    }
  }

  // 2. Ensure no subtopic has the same name as any main topic.
  for (const topic of response) {
    const subtopicNames =
      "subtopics" in topic ? topic.subtopics.map((subtopic) => subtopic.name) : [];
    for (const subtopicName of subtopicNames) {
      if (topicNames.includes(subtopicName) && subtopicName !== "Other") {
        console.warn(
          `Invalid response: Subtopic "${subtopicName}" has the same name as a main topic.`
        );
        return false;
      }
    }
  }

  return true;
}
